// ngx-social-login/esm2015/lib/providers/facebook-provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { OauthProvider } from './oauth-provider';
import { Observable } from 'rxjs';
import { Provider } from '../models/provider';
export class FacebookProvider extends OauthProvider {
    /**
     * @param {?} _config
     */
    constructor(_config) {
        super(Provider.FACEBOOK, '//connect.facebook.net/en_US/sdk.js', () => {
            const /** @type {?} */ initConfig = this._config.initOptions;
            const /** @type {?} */ config = initConfig.version ? initConfig : Object.assign({}, initConfig, { version: 'v2.9' });
            FB.init(config);
        });
        this._config = _config;
    }
    /**
     * @return {?}
     */
    login() {
        return Observable.create(observer => {
            FB.login((response) => {
                if (response.authResponse) {
                    const /** @type {?} */ authResponse = response.authResponse;
                    FB.api('/me?fields=name,email,picture,first_name,last_name', (fbUser) => {
                        observer.complete(/** @type {?} */ ({
                            id: fbUser.id,
                            name: fbUser.name,
                            email: fbUser.email,
                            profileImg: `https://graph.facebook.com/${fbUser.id}/picture?type=normal`,
                            firstName: fbUser.first_name,
                            lastName: fbUser.last_name,
                            accessToken: authResponse.accessToken
                        }));
                    });
                }
            }, this._config.loginOptions);
        });
    }
    /**
     * @return {?}
     */
    logout() {
        return Observable.create(observer => observer.complete(FB.logout()));
    }
}
function FacebookProvider_tsickle_Closure_declarations() {
    /** @type {?} */
    FacebookProvider.prototype._config;
}
