// ngx-social-login/esm2015/lib/providers/google-provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Observable, from } from 'rxjs';
import { OauthProvider } from './oauth-provider';
import { map } from 'rxjs/operators';
import { Provider } from '../models/provider';
export class GoogleProvider extends OauthProvider {
    /**
     * @param {?} config
     */
    constructor(config) {
        super(Provider.GOOGLE, '//apis.google.com/js/platform.js', () => {
            gapi.load('auth2', () => {
                const /** @type {?} */ conf = !config.scope ? Object.assign({}, config, { scope: 'email' }) : config;
                gapi.auth2
                    .init(conf)
                    .then(auth => (this._googleAuth = auth))
                    .catch(err => {
                    throw new Error(`Something went wrong during initialization: ${err.details}`);
                });
            });
        });
    }
    /**
     * @return {?}
     */
    login() {
        return from(this._googleAuth.signIn({ prompt: 'select_account' })).pipe(map(() => {
            const /** @type {?} */ profile = this._googleAuth.currentUser.get().getBasicProfile();
            const /** @type {?} */ accessToken = this._googleAuth.currentUser.get().getAuthResponse(true).access_token;
            const /** @type {?} */ idToken = this._googleAuth.currentUser.get().getAuthResponse(true).id_token;
            return {
                id: profile.getId(),
                email: profile.getEmail(),
                name: profile.getName(),
                profileImg: profile.getImageUrl(),
                accessToken: accessToken,
                idToken: idToken
            };
        }));
    }
    /**
     * @return {?}
     */
    logout() {
        return Observable.create(observer => observer.complete(this._googleAuth.signOut()));
    }
}
function GoogleProvider_tsickle_Closure_declarations() {
    /** @type {?} */
    GoogleProvider.prototype._googleAuth;
}
