// ngx-social-login/fesm2015/ngx-social-login.js
import { InjectionToken, Inject, Injectable, NgModule } from '@angular/core';
import { Observable, from, EMPTY } from 'rxjs';
import { map } from 'rxjs/operators';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const /** @type {?} */ CONFIG = new InjectionToken('providers configuration');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @abstract
 */
class OauthProvider {
    /**
     * @param {?} provider
     * @param {?} src
     * @param {?} onload
     * @param {?=} async
     */
    constructor(provider, src, onload, async = true) {
        this.ID_PREFIX = 'social-login__';
        const /** @type {?} */ id = this.ID_PREFIX + provider;
        if (document.getElementById(id)) {
            return;
        }
        const /** @type {?} */ script = document.createElement('script');
        script.id = id;
        script.async = async;
        script.src = src;
        script.onload = onload;
        script.defer = true;
        document.head.appendChild(script);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {string} */
const Provider = {
    FACEBOOK: 'facebook',
    GOOGLE: 'google',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class FacebookProvider extends OauthProvider {
    /**
     * @param {?} _config
     */
    constructor(_config) {
        super(Provider.FACEBOOK, '//connect.facebook.net/en_US/sdk.js', () => {
            const /** @type {?} */ initConfig = this._config.initOptions;
            const /** @type {?} */ config = initConfig.version ? initConfig : Object.assign({}, initConfig, { version: 'v2.9' });
            FB.init(config);
        });
        this._config = _config;
    }
    /**
     * @return {?}
     */
    login() {
        return Observable.create(observer => {
            FB.login((response) => {
                if (response.authResponse) {
                    const /** @type {?} */ authResponse = response.authResponse;
                    FB.api('/me?fields=name,email,picture,first_name,last_name', (fbUser) => {
                        observer.complete(/** @type {?} */ ({
                            id: fbUser.id,
                            name: fbUser.name,
                            email: fbUser.email,
                            profileImg: `https://graph.facebook.com/${fbUser.id}/picture?type=normal`,
                            firstName: fbUser.first_name,
                            lastName: fbUser.last_name,
                            accessToken: authResponse.accessToken
                        }));
                    });
                }
            }, this._config.loginOptions);
        });
    }
    /**
     * @return {?}
     */
    logout() {
        return Observable.create(observer => observer.complete(FB.logout()));
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class GoogleProvider extends OauthProvider {
    /**
     * @param {?} config
     */
    constructor(config) {
        super(Provider.GOOGLE, '//apis.google.com/js/platform.js', () => {
            gapi.load('auth2', () => {
                const /** @type {?} */ conf = !config.scope ? Object.assign({}, config, { scope: 'email' }) : config;
                gapi.auth2
                    .init(conf)
                    .then(auth => (this._googleAuth = auth))
                    .catch(err => {
                    throw new Error(`Something went wrong during initialization: ${err.details}`);
                });
            });
        });
    }
    /**
     * @return {?}
     */
    login() {
        return from(this._googleAuth.signIn({ prompt: 'select_account' })).pipe(map(() => {
            const /** @type {?} */ profile = this._googleAuth.currentUser.get().getBasicProfile();
            const /** @type {?} */ accessToken = this._googleAuth.currentUser.get().getAuthResponse(true).access_token;
            const /** @type {?} */ idToken = this._googleAuth.currentUser.get().getAuthResponse(true).id_token;
            return {
                id: profile.getId(),
                email: profile.getEmail(),
                name: profile.getName(),
                profileImg: profile.getImageUrl(),
                accessToken: accessToken,
                idToken: idToken
            };
        }));
    }
    /**
     * @return {?}
     */
    logout() {
        return Observable.create(observer => observer.complete(this._googleAuth.signOut()));
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class SocialLoginService {
    /**
     * @param {?} config
     */
    constructor(config) {
        this._providers = {};
        this._providers = Object.keys(config).reduce((obj, providerId) => {
            const /** @type {?} */ provider = Provider[providerId.toUpperCase()];
            obj[provider] = this.oauthProviderFactory(provider, config[providerId]);
            return obj;
        }, {});
    }
    /**
     * @param {?} provider
     * @return {?}
     */
    login(provider) {
        this._providerInUse = provider;
        const /** @type {?} */ oauthProvider = this._providers[provider];
        return oauthProvider ? oauthProvider.login() : EMPTY;
    }
    /**
     * @return {?}
     */
    logout() {
        const /** @type {?} */ oauthProvider = this._providers[this._providerInUse];
        return oauthProvider ? oauthProvider.logout() : EMPTY;
    }
    /**
     * @param {?} provider
     * @param {?} config
     * @return {?}
     */
    oauthProviderFactory(provider, config) {
        switch (provider) {
            case Provider.FACEBOOK:
                return new FacebookProvider(/** @type {?} */ (config));
            case Provider.GOOGLE:
                return new GoogleProvider(/** @type {?} */ (config));
        }
    }
}
SocialLoginService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
SocialLoginService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [CONFIG,] }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NgxSocialLoginModule {
    /**
     * @param {?} config
     * @return {?}
     */
    static init(config) {
        return {
            ngModule: NgxSocialLoginModule,
            providers: [
                { provide: CONFIG, useValue: config }
            ]
        };
    }
}
NgxSocialLoginModule.decorators = [
    { type: NgModule, args: [{
                providers: [SocialLoginService]
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { SocialLoginService, NgxSocialLoginModule, Provider, CONFIG as ɵa };
