// ngx-social-login/fesm5/ngx-social-login.js
import { InjectionToken, Inject, Injectable, NgModule } from '@angular/core';
import { __extends, __assign } from 'tslib';
import { Observable, from, EMPTY } from 'rxjs';
import { map } from 'rxjs/operators';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var /** @type {?} */ CONFIG = new InjectionToken('providers configuration');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
OauthProvider = /** @class */ (function () {
    function OauthProvider(provider, src, onload, async) {
        if (async === void 0) { async = true; }
        this.ID_PREFIX = 'social-login__';
        var /** @type {?} */ id = this.ID_PREFIX + provider;
        if (document.getElementById(id)) {
            return;
        }
        var /** @type {?} */ script = document.createElement('script');
        script.id = id;
        script.async = async;
        script.src = src;
        script.onload = onload;
        script.defer = true;
        document.head.appendChild(script);
    }
    return OauthProvider;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {string} */
var Provider = {
    FACEBOOK: 'facebook',
    GOOGLE: 'google',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var FacebookProvider = /** @class */ (function (_super) {
    __extends(FacebookProvider, _super);
    function FacebookProvider(_config) {
        var _this = _super.call(this, Provider.FACEBOOK, '//connect.facebook.net/en_US/sdk.js', function () {
            var /** @type {?} */ initConfig = _this._config.initOptions;
            var /** @type {?} */ config = initConfig.version ? initConfig : __assign({}, initConfig, { version: 'v2.9' });
            FB.init(config);
        }) || this;
        _this._config = _config;
        return _this;
    }
    /**
     * @return {?}
     */
    FacebookProvider.prototype.login = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Observable.create(function (observer) {
            FB.login(function (response) {
                if (response.authResponse) {
                    var /** @type {?} */ authResponse_1 = response.authResponse;
                    FB.api('/me?fields=name,email,picture,first_name,last_name', function (fbUser) {
                        observer.complete(/** @type {?} */ ({
                            id: fbUser.id,
                            name: fbUser.name,
                            email: fbUser.email,
                            profileImg: "https://graph.facebook.com/" + fbUser.id + "/picture?type=normal",
                            firstName: fbUser.first_name,
                            lastName: fbUser.last_name,
                            accessToken: authResponse_1.accessToken
                        }));
                    });
                }
            }, _this._config.loginOptions);
        });
    };
    /**
     * @return {?}
     */
    FacebookProvider.prototype.logout = /**
     * @return {?}
     */
    function () {
        return Observable.create(function (observer) { return observer.complete(FB.logout()); });
    };
    return FacebookProvider;
}(OauthProvider));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var GoogleProvider = /** @class */ (function (_super) {
    __extends(GoogleProvider, _super);
    function GoogleProvider(config) {
        var _this = _super.call(this, Provider.GOOGLE, '//apis.google.com/js/platform.js', function () {
            gapi.load('auth2', function () {
                var /** @type {?} */ conf = !config.scope ? __assign({}, config, { scope: 'email' }) : config;
                gapi.auth2
                    .init(conf)
                    .then(function (auth) { return (_this._googleAuth = auth); })
                    .catch(function (err) {
                    throw new Error("Something went wrong during initialization: " + err.details);
                });
            });
        }) || this;
        return _this;
    }
    /**
     * @return {?}
     */
    GoogleProvider.prototype.login = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return from(this._googleAuth.signIn({ prompt: 'select_account' })).pipe(map(function () {
            var /** @type {?} */ profile = _this._googleAuth.currentUser.get().getBasicProfile();
            var /** @type {?} */ accessToken = _this._googleAuth.currentUser.get().getAuthResponse(true).access_token;
            var /** @type {?} */ idToken = _this._googleAuth.currentUser.get().getAuthResponse(true).id_token;
            return {
                id: profile.getId(),
                email: profile.getEmail(),
                name: profile.getName(),
                profileImg: profile.getImageUrl(),
                accessToken: accessToken,
                idToken: idToken
            };
        }));
    };
    /**
     * @return {?}
     */
    GoogleProvider.prototype.logout = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Observable.create(function (observer) { return observer.complete(_this._googleAuth.signOut()); });
    };
    return GoogleProvider;
}(OauthProvider));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SocialLoginService = /** @class */ (function () {
    function SocialLoginService(config) {
        var _this = this;
        this._providers = {};
        this._providers = Object.keys(config).reduce(function (obj, providerId) {
            var /** @type {?} */ provider = Provider[providerId.toUpperCase()];
            obj[provider] = _this.oauthProviderFactory(provider, config[providerId]);
            return obj;
        }, {});
    }
    /**
     * @param {?} provider
     * @return {?}
     */
    SocialLoginService.prototype.login = /**
     * @param {?} provider
     * @return {?}
     */
    function (provider) {
        this._providerInUse = provider;
        var /** @type {?} */ oauthProvider = this._providers[provider];
        return oauthProvider ? oauthProvider.login() : EMPTY;
    };
    /**
     * @return {?}
     */
    SocialLoginService.prototype.logout = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ oauthProvider = this._providers[this._providerInUse];
        return oauthProvider ? oauthProvider.logout() : EMPTY;
    };
    /**
     * @param {?} provider
     * @param {?} config
     * @return {?}
     */
    SocialLoginService.prototype.oauthProviderFactory = /**
     * @param {?} provider
     * @param {?} config
     * @return {?}
     */
    function (provider, config) {
        switch (provider) {
            case Provider.FACEBOOK:
                return new FacebookProvider(/** @type {?} */ (config));
            case Provider.GOOGLE:
                return new GoogleProvider(/** @type {?} */ (config));
        }
    };
    SocialLoginService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    SocialLoginService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [CONFIG,] }] }
    ]; };
    return SocialLoginService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var NgxSocialLoginModule = /** @class */ (function () {
    function NgxSocialLoginModule() {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    NgxSocialLoginModule.init = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: NgxSocialLoginModule,
            providers: [
                { provide: CONFIG, useValue: config }
            ]
        };
    };
    NgxSocialLoginModule.decorators = [
        { type: NgModule, args: [{
                    providers: [SocialLoginService]
                },] },
    ];
    return NgxSocialLoginModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { SocialLoginService, NgxSocialLoginModule, Provider, CONFIG as ɵa };
