// ngx-social-login/esm5/lib/providers/facebook-provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { OauthProvider } from './oauth-provider';
import { Observable } from 'rxjs';
import { Provider } from '../models/provider';
var FacebookProvider = /** @class */ (function (_super) {
    tslib_1.__extends(FacebookProvider, _super);
    function FacebookProvider(_config) {
        var _this = _super.call(this, Provider.FACEBOOK, '//connect.facebook.net/en_US/sdk.js', function () {
            var /** @type {?} */ initConfig = _this._config.initOptions;
            var /** @type {?} */ config = initConfig.version ? initConfig : tslib_1.__assign({}, initConfig, { version: 'v2.9' });
            FB.init(config);
        }) || this;
        _this._config = _config;
        return _this;
    }
    /**
     * @return {?}
     */
    FacebookProvider.prototype.login = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Observable.create(function (observer) {
            FB.login(function (response) {
                if (response.authResponse) {
                    var /** @type {?} */ authResponse_1 = response.authResponse;
                    FB.api('/me?fields=name,email,picture,first_name,last_name', function (fbUser) {
                        observer.complete(/** @type {?} */ ({
                            id: fbUser.id,
                            name: fbUser.name,
                            email: fbUser.email,
                            profileImg: "https://graph.facebook.com/" + fbUser.id + "/picture?type=normal",
                            firstName: fbUser.first_name,
                            lastName: fbUser.last_name,
                            accessToken: authResponse_1.accessToken
                        }));
                    });
                }
            }, _this._config.loginOptions);
        });
    };
    /**
     * @return {?}
     */
    FacebookProvider.prototype.logout = /**
     * @return {?}
     */
    function () {
        return Observable.create(function (observer) { return observer.complete(FB.logout()); });
    };
    return FacebookProvider;
}(OauthProvider));
export { FacebookProvider };
function FacebookProvider_tsickle_Closure_declarations() {
    /** @type {?} */
    FacebookProvider.prototype._config;
}
