// ngx-social-login/esm5/lib/providers/google-provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Observable, from } from 'rxjs';
import { OauthProvider } from './oauth-provider';
import { map } from 'rxjs/operators';
import { Provider } from '../models/provider';
var GoogleProvider = /** @class */ (function (_super) {
    tslib_1.__extends(GoogleProvider, _super);
    function GoogleProvider(config) {
        var _this = _super.call(this, Provider.GOOGLE, '//apis.google.com/js/platform.js', function () {
            gapi.load('auth2', function () {
                var /** @type {?} */ conf = !config.scope ? tslib_1.__assign({}, config, { scope: 'email' }) : config;
                gapi.auth2
                    .init(conf)
                    .then(function (auth) { return (_this._googleAuth = auth); })
                    .catch(function (err) {
                    throw new Error("Something went wrong during initialization: " + err.details);
                });
            });
        }) || this;
        return _this;
    }
    /**
     * @return {?}
     */
    GoogleProvider.prototype.login = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return from(this._googleAuth.signIn({ prompt: 'select_account' })).pipe(map(function () {
            var /** @type {?} */ profile = _this._googleAuth.currentUser.get().getBasicProfile();
            var /** @type {?} */ accessToken = _this._googleAuth.currentUser.get().getAuthResponse(true).access_token;
            var /** @type {?} */ idToken = _this._googleAuth.currentUser.get().getAuthResponse(true).id_token;
            return {
                id: profile.getId(),
                email: profile.getEmail(),
                name: profile.getName(),
                profileImg: profile.getImageUrl(),
                accessToken: accessToken,
                idToken: idToken
            };
        }));
    };
    /**
     * @return {?}
     */
    GoogleProvider.prototype.logout = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Observable.create(function (observer) { return observer.complete(_this._googleAuth.signOut()); });
    };
    return GoogleProvider;
}(OauthProvider));
export { GoogleProvider };
function GoogleProvider_tsickle_Closure_declarations() {
    /** @type {?} */
    GoogleProvider.prototype._googleAuth;
}
